src/uct/ib/mlx5/ib_mlx5.c

```diff
@@ -405,8 +405,16 @@ ucs_status_t uct_ib_mlx5_devx_uar_init(uct_ib_mlx5_devx_uar_t *uar,
                                        uct_ib_mlx5_mmio_mode_t mmio_mode)
 {
 #if HAVE_DEVX
-    uar->uar            = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context, 0);
+    unsigned flags      = UCT_IB_MLX5_UAR_ALLOC_TYPE_WC;
+
+retry:
+    uar->uar            = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context,
+                                                flags);
     if (uar->uar == NULL) {
+        if (flags != UCT_IB_MLX5_UAR_ALLOC_TYPE_NC) {
+            flags = UCT_IB_MLX5_UAR_ALLOC_TYPE_NC;
+            goto retry;
+        }
```

Comment:
IMO, calling mlx5dv_devx_alloc_uar twice(instead of goto/retry) would be more readable

Reply:
don't we want to cache the failure?